### ISSUE TITLE
After updating to DOSBox Pure 1.0 Preview 1, none of my games work

### ISSUE THREAD

**Issue by SiriusVI87** [schellingb/dosbox-pure#566]
After updating to DOSBox Pure 1.0 Preview 1, none of my games work anymore. They worked perfectly with the previous version. My shader preset intro movie plays and then the screen stays black afterwards. I notice that I get a strange scanline / moire effect that I didn't get before either, so I'm guessing something is up with the graphics settings?

I'm launching my games via .conf file, such as this from 1st Person Pinball:

[sdl]
fullscreen=false
fulldouble=false
fullresolution=0x0
windowresolution=1280x960
output=openglnb
mapperfile=mapper.map
[dosbox]
machine=svga_s3
memsize=16
[render]
aspect=true
scaler=normal2x
[cpu]
core=auto
cputype=auto
cycles=800
[mixer]
[midi]
[sblaster]
[gus]
[speaker]
[joystick]
[serial]
[dos]
[ipx]
[autoexec]
mount c .\DOS\1stPerso
c:
@cd 1stPerso
@cls
@game
exit

**Comment by schellingb** [comment]
Can you toggle through the 5 settings under `Core Options` -> `Video` -> `Aspect Ratio Correction` and see if that affects anything while the screen has a weird pattern or is black? Then next can you try to run the same .conf file just without the line `aspect=true` and see if that changes things? Last, can you enable the RetroArch setting `User Interface` -> `On-Screen Notifications` -> `Notification Visibility` -> `Display Statistics` and take a screenshot of the statistics while the screen has a weird pattern or is black?

**Comment by SiriusVI87** [comment]
> Can you toggle through the 5 settings under `Core Options` -> `Video` -> `Aspect Ratio Correction` and see if that affects anything while the screen has a weird pattern or is black? Then next can you try to run the same .conf file just without the line `aspect=true` and see if that changes things? Last, can you enable the RetroArch setting `User Interface` -> `On-Screen Notifications` -> `Notification Visibility` -> `Display Statistics` and take a screenshot of the statistics while the screen has a weird pattern or is black?

Hey, thxfor writing back!

The first 2 suggestions did nothing unfortunately. This is tthe screenshot you asled for. Can you help me?

![image](https://github.com/user-attachments/assets/1e47a977-bc2e-462d-a86d-0fad2c3749d8)

**Comment by schellingb** [comment]
Certainly weird how it ends up with the resolution 0x0... Can you send a logfile? You might need to enable logging in RetroArch.

**Comment by SiriusVI87** [comment]
> Certainly weird how it ends up with the resolution 0x0... Can you send a logfile? You might need to enable logging in RetroArch.

Sure, I will. In the meantime here is the screenshot with the working version:

![image](https://github.com/user-attachments/assets/167f12fc-121c-40bd-9f63-85d42fa08c3f)

**Comment by SiriusVI87** [comment]
> Certainly weird how it ends up with the resolution 0x0... Can you send a logfile? You might need to enable logging in RetroArch.

This is my log file:

[retroarch.log](https://github.com/user-attachments/files/18261983/retroarch.log)

**Comment by schellingb** [comment]
Ah, the issue is that the line `mount` line is likely incorrect. In the old version the core incorrectly mounted relative paths (like `./something`) relative to a random directory, which can be the frontends directory but it can be something else depending on platform or previously launched cores. As mentioned in the release notes:

- Command-line mounting with relative paths were made relative to the loaded content (mount/imgmount) (59a20b1)

**Comment by SiriusVI87** [comment]
> Ah, the issue is that the line `mount` line is likely incorrect. In the old version the core incorrectly mounted relative paths (like `./something`) relative to a random directory, which can be the frontends directory but it can be something else depending on platform or previously launched cores. As mentioned in the release notes:
> 
> * Command-line mounting with relative paths were made relative to the loaded content (mount/imgmount) ([59a20b1](https://github.com/schellingb/dosbox-pure/commit/59a20b1f6709208fe4af507fde75b6a03dedf890))
> 
> So your autoexec fails to mount anything, so all the following lines are failing with errors except then exit which gives you the black screen because DOS has been exited.

Hey, thx for clearing that up, although I don't believe I know how to fix this. For my setup I rely on the relative paths to work without changing the .conf files.

So in the example below, where should I move the corresponding .conf file to, in order for the relative paths to work?

.\eXoDOS\1stPerso

**Comment by SiriusVI87** [comment]
Update:

I have kind of gotten it to work, but it requires a VERY extensive restructuring of my setup and I also don't like the results very much. 

1. I have to create an extra folder in "M:\Launchbox\Emulators\RetroArch" --> "M:\Launchbox\Emulators\RetroArch\eXoConfigs". 2. Then I have to copy all the .conf files in this folder (which means win3x and ms dos will be mixed up together, making it hard to conveniently import them into Launchbox).
3. Then I have to create a symlink to eXoDOS in the eXoConfigs folder, so "M:\Launchbox\Emulators\RetroArch\eXoConfigs\eXoDOS" and "M:\Launchbox\Emulators\RetroArch\eXoConfigs\eXoWin3x"

Is there a simpler solution? I very much preferred how the older version handled the relative paths. That way I could just leave each .conf file in the same folder of the game files.

May I ask what the benefit is of the new way relative paths are handled and if there is a chance of this change being reverted?

Thx!

**Comment by schellingb** [comment]
Can you explain how relative paths were handled before? It was random, so it might work on one device but not on another. Also if you run a certain core before DOSBox Pure, relative paths would not work anymore, etc. Now it always works, the paths in a .conf are relative to the .conf, how is that not much clearer?

**Comment by SiriusVI87** [comment]
> Can you explain how relative paths were handled before? It was random, so it might work on one device but not on another. Also if you run a certain core before DOSBox Pure, relative paths would not work anymore, etc. Now it always works, the paths in a .conf are relative to the .conf, how is that not much clearer?

No I cannot really explain. I guess it worked for me, because I had the symlinks setup in M:\Launchbox\Emulators\RetroArch
![image](https://github.com/user-attachments/assets/1f004584-bbfd-44f6-bcee-8a125419f929)


That was actually the hardest part to figure out, but once I did, it never failed me, no matter if another core was opened before or whatnot. This setup was bullet proof.

The only thing that didn't work was the audio config, which you helped me with in my other issue thread (thx so much again!!!)

With the new way the dosbox pure core works, I have to do extensive reordering of the .conf files, as explained above. It might be more logical, but it unfortunately completely breaks my setup.

My last resort might be bmass editing the .conf files and turing all the relative paths into absolute paths, but if I ever change my HDD volume letter / move the games to a new drive, I have to repeat the mass edit.

Long story short: With the new way the relative paths are handled, no solution is as elegant as it was with the previous version =/

**Comment by schellingb** [comment]
So, I can explain how it worked before, but it's a bit complicated :-)

First, some necessary info. A running process (like RetroArch) on an operating system like Windows or Linux has what is called a "working directory". If a program opens a file like "hello.txt" without specifying a full path, the file will be opened relative to the working directory of the process.

So in previous versions of the core, relative paths used for example in a mount or imgmount command was just that, it was opened as is and thus the files and directories were opened relative to the working directory of the RetroArch process. If you launch RetroArch on Windows by double clicking retroarch.exe or a default shortcut, Windows will use RetroArch's directory as the working directory. A shortcut can override the working directory though. If running from the command line, the current directory in the command line will be the working directory. How does it work on Android? I have no idea. How on PS Vita? Switch? OpenDingux? It's impossible to factor in every platform RetroArch runs on, so it needs to be avoided.

Now, a program can freely change its working directory. With libretro, that means both the frontend itself and any core can freely change the working directory. For example, DOSBox-SVN is a core that changes the working directory to the directory of the loaded content. So if you were to run DOSBox-SVN before the old version of DOSBox Pure, your setup would also be broken. It's hard to say which other cores also do this.

Anyway, relying on the working directory is not a good idea and as described above should be avoided. Because we don't know when it changes. If the core were to set it at startup, what if the frontend decides to change it later?

In the end, if D:\DOSGAMES\MyGame.conf in the `[autoexec]` section says `MOUNT C MyGameDir`, it makes the most sense to have that `MyGameDir` refer to `D:\DOSGAMES\MyGameDir`. That is the only intuitive answer.

Keep in mind that this core originally didn't even have support for .conf files. It was just added because the other DOSBox based cores also have it so people kept asking for it. The other cores also seem to use relative paths in a .conf be relative to the .conf file itself, so with the current behavior it matches them, which feels like a good consistency.

It sucks that this change breaks setups like yours, but I hope you can understand why this change was necessary. It's better to have a defined behavior that is consistent across all platforms. So when someone sets up their games to work on Windows, they can just copy a game over onto another device and be sure that it still works.

**Comment by SiriusVI87** [comment]
I see, thx for the explanation. The only part I don't get is this one:

"Keep in mind that this core originally didn't even have support for .conf files. It was just added because the other DOSBox based cores also have it so people kept asking for it. The other cores also seem to use relative paths in a .conf be relative to the .conf file itself, so with the current behavior it matches them, which feels like a good consistency."

This does not seem to be the case, because my (broken) setup still works with the regular DOS BOX core (I haven't tried DOSBox-SVN).

So if DOSXBox Pure's behavior now matches the one in the other DOSBOX cores, how come my setup still works with these cores, but not with pure?

**Comment by SiriusVI87** [comment]
Curious, maybe the other cores were updated recently, too? I don't know. Anyway I will try some stuff, I have an idea on how to set up my, ehm ... setup so that it's not more complicated than before. I will report back once I get everything working and then we can close this issue.

Thx for your support! You have been a great help!

**Comment by SiriusVI87** [comment]
Hey, so I just changed my setup to something I thought worked, but apparently it only works with some games and I don't know why =(.

So with your explanation above, I found that I don't even need to have the .conf files in my Retroarch Directory, I can have them wherever if I create symlinks in such a way that the relative paths within the .conf files work.

So my new Setup is:

The game files are here in O:\LaunchBox\eXo\eXoWin3x

![image](https://github.com/user-attachments/assets/90d09e3b-1484-41d1-b215-fad6acdc5bb5)

The conf files are in O:\LaunchBox\Games\eXoConfigs\!win3x (that way I can have the conf files on the same drive as the game files all in one folder:

**Comment by schellingb** [comment]
What does not work? Try to remove the line "exit" so you can see what happens in the console, which of the commands fail etc.

**Comment by SiriusVI87** [comment]
> What does not work? Try to remove the line "exit" so you can see what happens in the console, which of the commands fail etc.

Same as before, black screen.

without the exit I get this:

![image](https://github.com/user-attachments/assets/15531677-be01-4b69-b145-b253e6191b91)

**Comment by SiriusVI87** [comment]
> Also remove the `cls`, we're trying to see the error message

This is what I got:

![image](https://github.com/user-attachments/assets/1d08a276-c076-4d0d-a8c7-b907686e638c)

**Comment by schellingb** [comment]
Now try reading the error message, doesn't this help you?

**Comment by SiriusVI87** [comment]
> Now try reading the error message, doesn't this help you?

No, I'm sorry. I don't get why it's not working for one game but it does for the other. And I don't get why it works with the regular dosbox core (seen my edited message above?)

I really don't get it. I toldyou was not super savvy. I just trial and error my way through here. =(

**Comment by SiriusVI87** [comment]
So I just tried my last resort option and made the path absolute:

[autoexec]
mount c O:\LaunchBox\eXo\eXoWin3x\3DUPCN\
imgmount d O:\LaunchBox\eXo\eXoWin3x\3DUPCN\cd\cd.cue -t iso
c: 
path=C:\;z:\;c:\windows\;c:\sierra\upball2\
@cd sierra\upball2

@win runexit pinball2.exe


But I still get the same weird error that somehow the path does not exist?

![image](https://github.com/user-attachments/assets/9ed9a437-cce9-495a-bc0e-7b6341ac6ac7)

But it's right there:

**Comment by schellingb** [comment]
So the absolute path doesn't even use the symbolic link anymore, right?

It certainly is weird that mount fails saying the directory doesn't exist but then imgmount succeeds mounting something that is INSIDE that directory. I will try to reproduce your setup with some dummy files using the exact same names on O: drive to see if I can get it happening.

**Comment by SiriusVI87** [comment]
> So the absolute path doesn't even use the symbolic link anymore, right?
> 
> It certainly is weird that mount fails saying the directory doesn't exist but then imgmount succeeds mounting something that is INSIDE that directory. I will try to reproduce your setup with some dummy files using the exact same names on O: drive to see if I can get it happening.

OMG, I got it:

This does not work:

[autoexec]
mount c .\eXoWin3x\3DUPCN\


This does:

[autoexec]
mount c .\eXoWin3x\3DUPCN\


However, to fix this, I would have to edit thousands of files manually, removing the last "\". I cannot think of any way to automate this.

Was this a desired change with the latest update? I SWEAR this worked in the last dosbox pure version and it also works with the regular dosbox core.

**Comment by schellingb** [comment]
Haha, I just realized that, too, it does not like a trailing `\` in the mount path. GitHub is certainly not helping with it hiding the `\` from your messages. I'd classify that as a bug, a path with a trailing `\` should still be a valid directory path. I'll fix this.

**Comment by SiriusVI87** [comment]
> Haha, I just realized that, too, it does not like a trailing `\` in the mount path. GitHub is certainly not helping with it hiding the `\` from your messages. I'd classify that as a bug, a path with a trailing `\` should still be a valid directory path. I'll fix this.
> 
> Because I copied your messages which have the `\` hidden in the mount command and it worked for me...

You can't imagine how happy I am that it's an actual bug and that it's an easy fix for you. THANK YOU for sticking with me. With this bug fixed I believe I can get a working setup that is probably even superior to my previous one. =)

Uff, I think I need a beer now.

**Comment by schellingb** [comment]
Thanks for sticking around and helping find this bug. I'll push this probably as an update to this version in a day or two if nothing else comes up. Here's a build of the core with the fix for 64-Bit Windows, if you can give it a go and check if the fix works, that would be great. [dosbox_pure_libretro_win64_dll_git_8e608c2.zip](https://github.com/user-attachments/files/18267113/dosbox_pure_libretro_win64_dll_git_8e608c2.zip)

**Comment by SiriusVI87** [comment]
> Thanks for sticking around and helping find this bug. I'll push this probably as an update to this version in a day or two if nothing else comes up. Here's a build of the core with the fix for 64-Bit Windows, if you can give it a go and check if the fix works, that would be great. [dosbox_pure_libretro_win64_dll_git_8e608c2.zip](https://github.com/user-attachments/files/18267113/dosbox_pure_libretro_win64_dll_git_8e608c2.zip)

Thx so much for giving me a preview version to try out. I just restructured my setup and it finally works in a way I am very happy with. =)

Thanks again for all your hard work and for guiding a noob through this process.

**Comment by SiriusVI87** [comment]
@schellingb Hey man, hope you are doing fine!

I recently updated to the preview 5 version of DOSBox Pure and it seems that the issue that was fixed in the "dosbox_pure_libretro_win64_dll_git_8e608c2"-version that you provided for me is back.

Anyway, I was just wondering if this is a regression and if this will eventually be integrated in a future update.

Thank you!

EDIT:

I quickly tested it myself:

Preview versions 2 and 3 work. They seem to have the fix that you made.

Preview versions 4 and 5 gives me the same behavior hat preview version 1 had back then.

EDIT 2:

It may be a different issue ... but I'm not sure. Let me know if I should open a separate ticket for this:

So with preview 5 THIS works:

```
[autoexec]
@mount c .\DOS-GAMES\dumast2
imgmount d .\DOS-GAMES\dumast2\cd\DM2.cue -t cdrom
@c:
cls
@call run
exit
```

The following does not work with preview 4 and 5, but it works with preview 2 and 3:

```
[autoexec]
cd ..
@cd ..
@mount c .\DOS-GAMES\
imgmount d .\DOS-GAMES\3SKULLS\cd\3skulls.cue -t cdrom
@c:
cls
@cd 3skulls
@call run
exit
```

The reason why I now think this is a different issue is that when I delete the backslash from like so:

```
@mount c .\DOS-GAMES
```

preview 5 still won't run it.

EDIT 3:

Some more testing --> THIS works with preview 5:

```
[autoexec]
cd ..
@cd ..
@mount c .\DOS-GAMES\3SKULLS
imgmount d .\DOS-GAMES\3SKULLS\cd\3skulls.cue -t cdrom
@c:
cls
@cd 3skulls
@call run
exit
```

I'm sure there is a good reason for this change, but it again breaks my set-up unfortunately and fixing this on my end would mean digging through and editing thousands of files. Do you have any suggestions on what I can do to make my existing setup function with the newest core versions of Dosbox Pure?

Thank you!
